src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -327,17 +328,130 @@ namespace OpenMS
+     *
+     * @param mz_rt_ranges A vector of pairs of RangeMZ and RangeRT objects representing the mz-rt ranges.
+     * @return A vector of pairs of size_t values representing the indices of the ranges in the MSExperiment object.
+     */
+    std::vector<std::pair<size_t,size_t>> getRangesIdcs_(const std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges) const;
+    
+    /**
+     * @brief returns start end end index of spectra that fall within the given rt range
+     */
+    std::pair<Size, Size> getSpectraIdxRangeByRetentionTime_(double start_rt, double end_rt) const;
+
+    /**
+     * @brief Retrieves the indices of spectra based on retention time.
+     *
+     * This function returns a vector of indices corresponding to spectra that fall within the specified retention time range and have the specified MS level.
+     *
+     * @param start The start of the retention time range.
+     * @param end The end of the retention time range.
+     * @param ms_level The MS level of the spectra to consider.
+     * @return A vector of indices of spectra that meet the specified criteria.
+     */
+    std::vector<Size> getSpectraIdcsByRetentionTime(double start, double end, unsigned int ms_level) const;
+
+    /**
+     * @brief Type alias for a function that aggregates data from a range of MSSpectrum objects.
+     *
+     * This function type is used as a parameter in various algorithms that operate on MSExperiment objects.
+     * The function takes a reference to an MSSpectrum object, as well as the start and end indices of a range of spectra,
+     * and returns a CoordinateType value that represents the aggregated result.
+     *
+     * @param s The MSSpectrum object to aggregate data from.
+     * @param start The peark start index.
+     * @param end The peak end index.
+     * @return The aggregated result as a CoordinateType value.
+     */
+    using AggregatorFunc = std::function<CoordinateType(const MSSpectrum& s, size_t start, size_t end)>;
+    
+    /**
+     * @brief Type alias for a reduce function.
+     *
+     * The reduce function is typically used to combine multiple values into a single value, such as in aggregation operations.
+     *
+     * @param arg1 The first argument of the reduce function.
+     * @param arg2 The second argument of the reduce function.
+     * @return The result of the reduce function.
+     */
+    using ReduceFunc = std::function<CoordinateType(CoordinateType&, CoordinateType)>;
+
+    /** @brief Aggregate all peaks in the range given by the begin and end area iterators.
+     *  The aggregation is performed in the order of the dimensions of the peaks.
+     *  Note: This allows to e.g., calculate total ion current.
+    **/ 
+    MSExperiment::CoordinateType aggregate(ConstAreaIterator begin, ConstAreaIterator end, AggregatorFunc rt, AggregatorFunc mz) const;
```

Comment:
@jpfeuffer there are quite a few variants. are all needed?

Reply:
Maybe we should still benchmark a reduce function instead of an aggregation.

In my initial tests aggregation was slightly faster.

Reply:
We should add back a function for the second dimension though

Reply:
the std::reduce should be faster but I did not get it to work with different types (double vs peak).

Yes we can add one. Did you have aggregation over the second dimension in mind? or more a transform +back_inserter or similar?

Reply:
Yes but in my tests (maybe because of std function), reduce was slower. Reduce is also more limited (no mean calculation for example).
Second dimension aggregation could be potentially Peak group detection, while the data is still hot.